README
CSS::DOM, version 0.15

This module implements a CSS-specific subset of the interfaces
described in the W3C DOM specification.


RECENT CHANGES

0.015
-----
• compute_style  no longer produces an  uninitialized  warning  under
  some circumstances (RT #79855).
• A minor backtracking bug in CSS::DOM::PropertyParser has been fixed.

0.014
-----
• The full text of the licence is now included  in  the  distribution.
  Thanks to Ville Skyttä for providing  it  (RT  #60770).
• A few pod syntax errors have been corrected, thanks to Nicholas Bam-
  ber (RT #63388).
• CSS::DOM::PropertyParser  has been fixed to work with  perl  5.13.8
  and higher.
• New getStringValue method of CSS::DOM::Value::Primitive

0.013
-----
Oops. Fix another problem, introduced in 0.12. I forgot to add a file
to MANIFEST. (I *thought* I had run make disttest.)


INSTALLATION

The easiest way to install this module is to use the CPAN module or
the cpan script:

    perl -MCPAN -e "install CSS::DOM"

    cpan CSS::DOM

Or you can use the following:

    perl Makefile.PL
    make
    make test
    [sudo] make install


DEPENDENCIES

This module requires perl 5.8.2 or later, and the following modules:

• Exporter 5.57 or later
• Encode 2.10 or later
• Clone 0.09 or later


DOCUMENTATION

After installing, you can find documentation for these modules with
the perldoc command.

    perldoc CSS::DOM

Or try using man (it's faster, in my experience):

    man CSS::DOM


COPYRIGHT AND LICENCE

Copyright (C) 2007-12 Father Chrysostomos

This program is free software; you may redistribute it and/or modify
it under the same terms as perl. The full text of the license can be
found in the LICENSE file included with this module.


Changes
Revision history for CSS::DOM

0.15    21 October, 2012
      • compute_style  no longer produces an uninitialized warning if
        the element’s own style attribute has a  property  for  which
        there is no applicable property with the  same  name  in  the
        style sheets (RT #79855).
      • A minor backtracking bug in CSS::DOM::PropertyParser has been
        fixed.  If a format for a list property can still match after
        backtracking past a capturing group,  then extra items  (from
        the pre-backtracking matches)  will end up in the list.  This
        can happen with  property  formats  like  ‘[(foo) | (foo),]+’.
        Any such format would  be  more  efficient  when  written  as
       ‘[(foo),?]+’, which does not exhibit the bug.

0.14    12 December, 2010
      • The full text of the licence is now included in the  distribu-
        tion.  Thanks to Ville Skyttä for providing  it  (RT  #60770).
      • A few pod syntax errors have been corrected,  thanks to Nicho-
        las Bamber (RT #63388).
      • CSS::DOM::PropertyParser  has been fixed to  work  with  perl
        5.13.8 and higher.
      • New getStringValue method of CSS::DOM::Value::Primitive

0.13    22 August, 2010
        Oops.  Fix another problem,  introduced in 0.12.  I forgot to
        add  a  file  to  MANIFEST.  (I  *thought*  I  had  run  make
        disttest.)

0.12    22 August, 2010
        Fix dependency problem introduced in 0.11. (I tried to remove
        the dependency on constant::lexical,  gave up and reverted my
        changes, but forgot to add it back to Makefile.PL.  Oh well.)

0.11    15 August, 2010
      • setStringValue  and  setFloatValue  are  now  listed  in
        CSS::DOM::Interface.
      • CSS::DOM::Style’s  opacity  property is  now  listed  in
        CSS::DOM::Interface. (This is a temporary workaround for
        a design flaw.)

0.10    28 March, 2010
      • New setFloatValue and setStringValue methods of CSS::DOM::Val-
        ue::Primitive
      • Bug fix: CSS::DOM::Value::Primitive’s cssText did not set the
        value when passed an argument if the  existing  value  was  a
        string and there was no existing serialisation  recorded  (as
        happens if you call the constructor directly  without  provid-
        ing the css argument).
      • Bug fix:  Sub-value objects of a list value  used  to  become
       ‘unowned’ if their type changed (via cssText assignment),  mak-
        ing further modification attempts fail.

0.09    21 February, 2010
      • The property parser no longer gets confused in cygwin’s  perl
        5.10.0@34065 (maint;  aka ee8a906) when an rgb() colour comes
        after something else in a property value.  This bug also  pro-
        duced warnings.
      • The property parser in cygwin’s perl no longer  gets  offsets
        muddled  up  when  a  shorthand  property’s  format  contains
        <colour> within a parenthetical group,  the  <colour>  is not
        the first thing in  the  group  (e.g.,  ‘(<ident> <colour>)’),
        and a value  assigned  has  a  colour  in  hex  format  (e.g.,
       ‘honey #bee’,  which would cause just ‘#bee’,  not ‘honey #bee’
        to be assigned to the subproperty).
      • The property parser used  to  produces  warnings  and  create
        CSS_CUSTOM value objects instead of ‘primitives’ when a short-
        hand property was assigned to that had named  properties  [RT
        #54809]. This only occurred in perl 5.10.0. It turns out that
        this perl version’s  $^N  is more buggy than I  realised  and
        it’s not just cygwin’s  maint  snapshot  that  needs  special
        treatment.
      • CSS::DOM::Style’s  length  method no longer dies if no proper-
        ties have been added [RT #54810].

0.08    22 November, 2009
      • Incompatible change:  CSS::DOM::Value’s and CSS::DOM::Value::
        Primitive’s constructors’ interfaces have changed.  They  now
        take hash-style arguments.
      • CSSPrimitiveValue  objects  for  colours  are  now  supported.
      • %CSS::DOM::Constants::SuffixToConst,  mapping  dimension  suf-
        fixes to constants
      • CSS::DOM::Value::Primitive  now implements the Rect  and  RGB-
        Color interfaces.
      • CSS::DOM::Value::List  has  been  added.  It  implements  the
        CSSValueList interface.
      • CSS::DOM::Value::Primitive no longer inherits from CSS::DOM::
        Value, but claims that it DOES it.
      • getPropertyCSSValue now works with list properties.
      • CSS::DOM::Style’s property_parser method
      • CSS::DOM::Value::Primitive’s primitiveType and  getFloatValue
        methods
      • CSS::DOM::Primitive’s and CSS::DOM::Value’s  cssText  methods
        are now writable.
      • Assigning to a shorthand property  (like border-top)  a value
        (like 'inset')  that causes some sub-properties to have their
        initial values now deletes any  implied  sub-properties  that
        do not have default values (like border-top-color).  (If that
        makes  no  sense  to  you,  see  the  tests  at  the  end  of
        property-parser.t.)
      • Fixed to be compatible with 5.10.0@34065 (maint; aka ee8a906),
        which is included with cygwin.

0.07    16 August, 2009
      • Support for specifications that describe  which  properties
        are supported and  also  the  syntax  of  shorthand  proper-
        ties  (CSS::DOM::PropertyParser  was  added;  CSS::DOM  and
        CSS::DOM::Style constructors now take more args)
      • Incompatible change:  CSS::DOM::Value is no longer used  with-
        out a property spec.
      • CSS::DOM has two new methods, url_fetcher and property_parser,
        which return what was passed to the constructor.
      • Bug fix  (broken in  0.06):  Assigning to a property an empty
        string or a string beginning with whitespace or a CSS comment
        no longer causes an error.

lib/CSS/DOM/Interface.pm
package CSS::DOM::Interface;

use Exporter 5.57 'import';
our $VERSION = '0.15';

require CSS::DOM::Constants;

=head1 NAME

CSS::DOM::Interface - A list of CSS::DOM's interface members in machine-readable format

=head1 VERSION

Version 0.15

=head1 SYNOPSIS

  use CSS::DOM::Interface ':all';

  # name of DOM interface (CSSStyleRule):
  $CSS::DOM::Interface{"CSS::DOM::Rule::Style"};
  
  # interface it inherits from (CSSRule):
  $CSS::DOM::Interface{CSSStyleRule}{_isa};
  
  # whether this can be used as an array
  $CSS::DOM::Interface{MediaList}{_array}; # true
  # or hash
  $CSS::DOM::Interface{MediaList}{_hash}; # false
  
  
  # Properties and Methods
  
  # list them all
  grep !/^_/, keys %{ $CSS::DOM::Interface{CSSStyleSheet} };
  
  # see whether a given property is supported
  exists $CSS::DOM::Interface{CSSStyleSheet}{foo}; # false
  
  # Is it a method?
  $CSS::DOM::Interface{CSSStyleSheet}{cssRules}   & METHOD; # false
  $CSS::DOM::Interface{CSSStyleSheet}{insertRule} & METHOD; # true
  
  # Does the method return nothing?
  $CSS::DOM::Interface{MediaList}{deleteMedium} & VOID; # true
  
  # Is a property read-only?
  $CSS::DOM::Interface{StyleSheetList}{length} & READONLY; # true
  
  # Data types of properties
  ($CSS::DOM::Interface{CSSStyleSheet}{type}      & TYPE)
                                                    == STR;  # true
  ($CSS::DOM::Interface{CSSStyleSheet}{disabled}  & TYPE)
                                                    == BOOL; # true
  ($CSS::DOM::Interface{CSSStyleSheet}{ownerNode} & TYPE)
                                                    == NUM;  # false
  ($CSS::DOM::Interface{CSSStyleSheet}{href}     & TYPE)
                                                    == OBJ;  # false
  
  # and return types of methods:
  ($CSS::DOM::Interface{MediaList}{item} & TYPE) == STR;  # true
  ($CSS::DOM::Interface{CSSMediaRule}
                          ->{insertRule} & TYPE) == BOOL; # false
  ($CSS::DOM::Interface{CSSStyleDeclaration}
                     ->{getPropertyVaue} & TYPE) == NUM;  # false
  ($CSS::DOM::Interface{CSSStyleDeclaration}
                      ->{removeProperty} & TYPE) == OBJ;  # false
  
  
  # Constants

  # list of constant names in the form "CSS::DOM::Node::STYLE_RULE";
  @{ $CSS::DOM::Interface{CSSRule}{_constants} };

=head1 DESCRIPTION

The synopsis should tell you almost everything you need to know. But be
warned that C<$foo & TYPE> is meaningless when C<$foo & METHOD> and
C<$foo & VOID> are both true. For more
gory details, look at the source code. In fact, here it is:

=cut

0 and q r

=for ;

  our @EXPORT_OK = qw/METHOD VOID READONLY BOOL STR NUM OBJ TYPE/;
  our %EXPORT_TAGS = (all => \@EXPORT_OK);

  sub METHOD   () {      1 }
  sub VOID     () {   0b10 } # for methods
  sub READONLY () {   0b10 } # for properties
  sub BOOL     () { 0b0000 }
  sub STR      () { 0b0100 }
  sub NUM      () { 0b1000 }
  sub OBJ      () { 0b1100 }
  sub TYPE     () { 0b1100 } # only for use as a mask

  %CSS::DOM::Interface = (
  	'CSS::DOM' => 'CSSStyleSheet',
  	'CSS::DOM::StyleSheetList' => 'StyleSheetList',
  	'CSS::DOM::MediaList' => 'MediaList',
  	'CSS::DOM::RuleList' => 'CSSRuleList',
  	'CSS::DOM::Rule' => 'CSSRule',
  	'CSS::DOM::Rule::Style' => 'CSSStyleRule',
  	'CSS::DOM::Rule::Media' => 'CSSMediaRule',
  	'CSS::DOM::Rule::FontFace' => 'CSSFontFaceRule',
  	'CSS::DOM::Rule::Page' => 'CSSPageRule',
  	'CSS::DOM::Rule::Import' => 'CSSImportRule',
  	'CSS::DOM::Rule::Charset' => 'CSSCharsetRule',
  	'CSS::DOM::Style' => 'CSSStyleDeclaration',
  	'CSS::DOM::Value' => 'CSSValue',
  	'CSS::DOM::Value::Primitive' => 'CSSPrimitiveValue',
  	'CSS::DOM::Value::List' => 'CSSValueList',
  	'CSS::DOM::Counter' => 'Counter',
  	 StyleSheetList => {
		_hash => 0,
		_array => 1,
  		length => NUM | READONLY,
  		item => METHOD | OBJ,
  	 },
  	 MediaList => {
		_hash => 0,
		_array => 1,
  		mediaText => STR,
  		length => NUM | READONLY,
  		item => METHOD | STR,
  		deleteMedium => METHOD | VOID,
  		appendMedium => METHOD | VOID,
  	 },
  	 CSSRuleList => {
		_hash => 0,
		_array => 1,
  		length => NUM | READONLY,
  		item => METHOD | OBJ,
  	 },
  	 CSSRule => {
		_hash => 0,
		_array => 0,
  		_constants => [qw[
  			CSS::DOM::Rule::UNKNOWN_RULE
  			CSS::DOM::Rule::STYLE_RULE
  			CSS::DOM::Rule::CHARSET_RULE
  			CSS::DOM::Rule::IMPORT_RULE
  			CSS::DOM::Rule::MEDIA_RULE
  			CSS::DOM::Rule::FONT_FACE_RULE
  			CSS::DOM::Rule::PAGE_RULE
  		]],
  		type => NUM | READONLY,
  		cssText => STR,
  		parentStyleSheet => OBJ | READONLY,
  		parentRule => OBJ | READONLY,
  	 },
  	 CSSStyleRule => {
		_isa => 'CSSRule',
		_hash => 0,
		_array => 0,
  		selectorText => STR,
  		style => OBJ | READONLY,
  	 },
  	 CSSMediaRule => {
		_isa => 'CSSRule',
		_hash => 0,
		_array => 0,
  		media => OBJ | READONLY,
  		cssRules => OBJ | READONLY,
  		insertRule => METHOD | NUM,
  		deleteRule => METHOD | VOID,
  	 },
  	 CSSFontFaceRule => {
		_isa => 'CSSRule',
		_hash => 0,
		_array => 0,
  		style => OBJ | READONLY,
  	 },
  	 CSSPageRule => {
		_isa => 'CSSRule',
		_hash => 0,
		_array => 0,
  		selectorText => STR,
  		style => OBJ | READONLY,
  	 },
  	 CSSImportRule => {
		_isa => 'CSSRule',
		_hash => 0,
		_array => 0,
  		href => STR | READONLY,
  		media => OBJ | READONLY,
  		styleSheet => OBJ | READONLY,
  	 },
  	 CSSCharsetRule => {
		_isa => 'CSSRule',
		_hash => 0,
		_array => 0,
  		encoding => STR,
  	 },
  	 CSSStyleDeclaration => {
		_hash => 0,
		_array => 1,
  		cssText => STR,
  		getPropertyValue => METHOD | STR,
  		getPropertyCSSValue => METHOD | OBJ,
  		removeProperty => METHOD | STR,
  		getPropertyPriority => METHOD | STR,
  		setProperty => METHOD | VOID,
  		length => NUM | READONLY,
  		item => METHOD | STR,
  		parentRule => OBJ | READONLY,
  		azimuth => STR,
  		background => STR,
  		backgroundAttachment => STR,
  		backgroundColor => STR,
  		backgroundImage => STR,
  		backgroundPosition => STR,
  		backgroundRepeat => STR,
  		border => STR,
  		borderCollapse => STR,
  		borderColor => STR,
  		borderSpacing => STR,
  		borderStyle => STR,
  		borderTop => STR,
  		borderRight => STR,
  		borderBottom => STR,
  		borderLeft => STR,
  		borderTopColor => STR,
  		borderRightColor => STR,
  		borderBottomColor => STR,
  		borderLeftColor => STR,
  		borderTopStyle => STR,
  		borderRightStyle => STR,
  		borderBottomStyle => STR,
  		borderLeftStyle => STR,
  		borderTopWidth => STR,
  		borderRightWidth => STR,
  		borderBottomWidth => STR,
  		borderLeftWidth => STR,
  		borderWidth => STR,
  		bottom => STR,
  		captionSide => STR,
  		clear => STR,
  		clip => STR,
  		color => STR,
  		content => STR,
  		counterIncrement => STR,
  		counterReset => STR,
  		cue => STR,
  		cueAfter => STR,
  		cueBefore => STR,
  		cursor => STR,
  		direction => STR,
  		display => STR,
  		elevation => STR,
  		emptyCells => STR,
  		cssFloat => STR,
  		font => STR,
  		fontFamily => STR,
  		fontSize => STR,
  		fontSizeAdjust => STR,
  		fontStretch => STR,
  		fontStyle => STR,
  		fontVariant => STR,
  		fontWeight => STR,
  		height => STR,
  		left => STR,
  		letterSpacing => STR,
  		lineHeight => STR,
  		listStyle => STR,
  		listStyleImage => STR,
  		listStylePosition => STR,
  		listStyleType => STR,
  		margin => STR,
  		marginTop => STR,
  		marginRight => STR,
  		marginBottom => STR,
  		marginLeft => STR,
  		markerOffset => STR,
  		marks => STR,
  		maxHeight => STR,
  		maxWidth => STR,
  		minHeight => STR,
  		minWidth => STR,
  		opacity => STR,
  		orphans => STR,
  		outline => STR,
  		outlineColor => STR,
  		outlineStyle => STR,
  		outlineWidth => STR,
  		overflow => STR,
  		padding => STR,
  		paddingTop => STR,
  		paddingRight => STR,
  		paddingBottom => STR,
  		paddingLeft => STR,
  		page => STR,
  		pageBreakAfter => STR,
  		pageBreakBefore => STR,
  		pageBreakInside => STR,
  		pause => STR,
  		pauseAfter => STR,
  		pauseBefore => STR,
  		pitch => STR,
  		pitchRange => STR,
  		playDuring => STR,
  		position => STR,
  		quotes => STR,
  		richness => STR,
  		right => STR,
  		size => STR,
  		speak => STR,
  		speakHeader => STR,
  		speakNumeral => STR,
  		speakPunctuation => STR,
  		speechRate => STR,
  		stress => STR,
  		tableLayout => STR,
  		textAlign => STR,
  		textDecoration => STR,
  		textIndent => STR,
  		textShadow => STR,
  		textTransform => STR,
  		top => STR,
  		unicodeBidi => STR,
  		verticalAlign => STR,
  		visibility => STR,
  		voiceFamily => STR,
  		volume => STR,
  		whiteSpace => STR,
  		widows => STR,
  		width => STR,
  		wordSpacing => STR,
  		zIndex => STR,
  	 },
  	 CSSValue => {
		_hash => 0,
		_array => 0,
  		_constants => [qw[
  			CSS::DOM::Value::CSS_INHERIT
  			CSS::DOM::Value::CSS_PRIMITIVE_VALUE
  			CSS::DOM::Value::CSS_VALUE_LIST
  			CSS::DOM::Value::CSS_CUSTOM
  		]],
  		cssText => STR,
  		cssValueType => NUM | READONLY,
  	 },
  	 CSSPrimitiveValue => {
		_isa => 'CSSValue',
		_hash => 0,
		_array => 0,
  		_constants => [qw[
  			CSS::DOM::Value::Primitive::CSS_UNKNOWN
  			CSS::DOM::Value::Primitive::CSS_NUMBER
  			CSS::DOM::Value::Primitive::CSS_PERCENTAGE
  			CSS::DOM::Value::Primitive::CSS_EMS
  			CSS::DOM::Value::Primitive::CSS_EXS
  			CSS::DOM::Value::Primitive::CSS_PX
  			CSS::DOM::Value::Primitive::CSS_CM
  			CSS::DOM::Value::Primitive::CSS_MM
  			CSS::DOM::Value::Primitive::CSS_IN
  			CSS::DOM::Value::Primitive::CSS_PT
  			CSS::DOM::Value::Primitive::CSS_PC
  			CSS::DOM::Value::Primitive::CSS_DEG
  			CSS::DOM::Value::Primitive::CSS_RAD
  			CSS::DOM::Value::Primitive::CSS_GRAD
  			CSS::DOM::Value::Primitive::CSS_MS
  			CSS::DOM::Value::Primitive::CSS_S
  			CSS::DOM::Value::Primitive::CSS_HZ
  			CSS::DOM::Value::Primitive::CSS_KHZ
  			CSS::DOM::Value::Primitive::CSS_DIMENSION
  			CSS::DOM::Value::Primitive::CSS_STRING
  			CSS::DOM::Value::Primitive::CSS_URI
  			CSS::DOM::Value::Primitive::CSS_IDENT
  			CSS::DOM::Value::Primitive::CSS_ATTR
  			CSS::DOM::Value::Primitive::CSS_COUNTER
  			CSS::DOM::Value::Primitive::CSS_RECT
  			CSS::DOM::Value::Primitive::CSS_RGBCOLOR
  		]],
  		primitiveType => NUM | READONLY,
  		setFloatValue => METHOD | VOID,
  		getFloatValue => METHOD | NUM,
  		setStringValue => METHOD | VOID,
  		getStringValue => METHOD | STR,
  #		getCounterValue => METHOD | OBJ,
  #		getRectValue => METHOD | OBJ,
  #		getRGBColorValue => METHOD | OBJ,
  		red => OBJ | READONLY,
  		green => OBJ | READONLY,
  		blue => OBJ | READONLY,
  		alpha => OBJ | READONLY,
  		top => OBJ | READONLY,
  		right => OBJ | READONLY,
  		bottom => OBJ | READONLY,
  		left => OBJ | READONLY,
  	 },
  	 CSSValueList => {
		_isa => 'CSSValue',
		_hash => 0,
		_array => 1,
  		length => NUM | READONLY,
  		item => METHOD | OBJ,
  	 },
  #	 Counter => {
  #		_hash => 0,
  #		_array => 0,
  #		identifier => STR | READONLY,
  #		listStyle => STR | READONLY,
  #		separator => STR | READONLY,
  #	 },
  	 CSSStyleSheet => {
  		type => STR | READONLY,
		_hash => 0,
		_array => 0,
  		disabled => BOOL,
  		ownerNode => OBJ | READONLY,
  		parentStyleSheet => OBJ | READONLY,
  		href => STR | READONLY,
  		title => STR | READONLY,
  		media => OBJ | READONLY,
  		ownerRule => OBJ | READONLY,
  		cssRules => OBJ | READONLY,
  		insertRule => METHOD | NUM,
  		deleteRule => METHOD | VOID,
  	 },
  );

__END__

=head1 SEE ALSO

L<CSS::DOM>
